File: App/Resources/AlgorithmDetails/foldsort/foldsort.rb
def comp_swap(arr, a, b, sort_end)
  if b < sort_end && arr[a] > arr[b]
    arr[a], arr[b] = arr[b], arr[a]
  end
end

def halver(arr, low, high, sort_end)
  while low < high
    comp_swap(arr, low, high, sort_end)
    low += 1
    high -= 1
  end
end

def sort(arr)
  n = arr.length
  ceil_log = 1
  while (1 << ceil_log) < n
    ceil_log += 1
  end
  sort_end = n
  size2 = 1 << ceil_log

  k = size2 >> 1
  while k > 0
    i = size2
    while i >= k
      j = 0
      while j < sort_end
        halver(arr, j, j + i - 1, sort_end)
        j += i
      end
      i >>= 1
    end
    k >>= 1
  end
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array

File: App/Resources/AlgorithmDetails/foldsort/foldsort.go
package main

import (
	"fmt"
)

func compSwap(arr []int, a, b, end int) {
	if b < end && arr[a] > arr[b] {
		arr[a], arr[b] = arr[b], arr[a]
	}
}

func halver(arr []int, low, high, end int) {
	for low < high {
		compSwap(arr, low, high, end)
		low++
		high--
	}
}

func sort(arr []int) []int {
	n := len(arr)
	ceilLog := 1
	for (1 << ceilLog) < n {
		ceilLog++
	}
	end := n
	size2 := 1 << ceilLog

	k := size2 >> 1
	for k > 0 {
		i := size2
		for i >= k {
			j := 0
			for j < end {
				halver(arr, j, j+i-1, end)
				j += i
			}
			i >>= 1
		}
		k >>= 1
	}
	return arr
}

func main() {
	array := []int{0, 39, 21, 62, 91, 77, 14, 23,
		90, 69, 51, 81, 68, 83, 32, 56}
	fmt.Println(sort(array))
}
